pkg/controller/serviceexternalip/controller.go

```diff
@@ -0,0 +1,391 @@
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing Service for %s. (%v)", key, time.Since(startTime))
+	}()
+
+	service, err := c.serviceLister.Services(key.Namespace).Get(key.Name)
+	if err != nil {
+		// Service already deleted
+		if apimachineryerrors.IsNotFound(err) {
+			if err := c.releaseExternalIP(key); err != nil {
+				return err
+			}
+			return nil
+		}
+		return err
+	}
+
+	// Service does not need external IP or type has changed.
+	if service.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		if err := c.releaseExternalIP(key); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	currentIPPool := service.ObjectMeta.Annotations[antreaagenttypes.ServiceExternalIPPoolAnnotationKey]
+	prevIPAllocation, allocationExists := c.getExternalIPAllocation(key)
+	currentExternalIP := getServiceExternalIP(service)
+
+	// If user specifies external IP in spec, we should check whether it matches the current external IP.
+	specIPMatched := service.Spec.LoadBalancerIP == "" || service.Spec.LoadBalancerIP == currentExternalIP
+
+	if allocationExists && specIPMatched &&
+		c.externalIPAllocator.IPPoolExists(currentIPPool) &&
+		c.externalIPAllocator.IPPoolHasIP(currentIPPool, prevIPAllocation.ip) &&
+		currentIPPool == prevIPAllocation.ipPool &&
+		currentExternalIP == prevIPAllocation.ip.String() {
+		return nil
+	}
+
+	// The ExternalIPPool does not exist or has been deleted. Reclaim the external IP.
+	if currentIPPool != "" && !c.externalIPAllocator.IPPoolExists(currentIPPool) {
+		if err := c.releaseExternalIP(key); err != nil {
+			return err
+		}
+		if currentExternalIP != "" {
+			toUpdate := service.DeepCopy()
+			toUpdate.Status.LoadBalancer.Ingress = nil
+			return c.updateServiceStatus(service, toUpdate)
+		}
+	}
+
+	// the external IP or ExternalIPPool changed somehow. Delete the previous allocation.
+	if err := c.releaseExternalIP(key); err != nil {
+		return err
+	}
+
+	if currentIPPool == "" {
+		klog.ErrorS(fmt.Errorf("annotation %q is required", antreaagenttypes.ServiceExternalIPPoolAnnotationKey),
+			"allocate external IP failed", "service", key,
+		)
+		return nil
+	}
+
+	var newExternalIP net.IP
+	if service.Spec.LoadBalancerIP != "" {
+		// check whether the specified IP is in the desired IP pool or not.
+		newExternalIP = net.ParseIP(service.Spec.LoadBalancerIP)
+		if !c.externalIPAllocator.IPPoolHasIP(currentIPPool, newExternalIP) {
+			klog.Error("IP pool %q does not contains required external IP %q", currentIPPool, newExternalIP)
+			return nil
+		}
+		err = c.externalIPAllocator.UpdateIPAllocation(currentIPPool, newExternalIP)
+	} else {
+		// Allocate IP from existing ExternalIPPool.
+		newExternalIP, err = c.externalIPAllocator.AllocateIPFromPool(currentIPPool)
+	}
+	if err != nil {
+		// If the ExternalIPPool does not exist, we can ignore the error sine the Service will get requeued by ExternalIPPool change events.
```

Comment:
```suggestion
		// If the ExternalIPPool does not exist, we can ignore the error since the Service will get requeued by ExternalIPPool change events.
```